src/tree.js

```diff
@@ -5,8 +5,96 @@ import * as Null from './null.js'
 import * as Reference from './reference.js'
 import * as Type from './type.js'
 
+import { base32 } from './reference.js'
+
 export { sha256 }
 
+export class StrictTraversal {
+  /**
+   * @param {object} source
+   * @returns {Iterable<[string, unknown]|[number, unknown]>}
+   */
+  static *entries(source) {
+    if (source instanceof Map) {
+      yield* source.entries()
+    } else if (source instanceof Set) {
+      for (const value of source.values()) {
+        yield [value, true]
+      }
+    } else if (Array.isArray(source)) {
+      yield* source.entries()
+    } else {
+      yield* Object.entries(source)
+    }
+  }
+}
+
+/**
+ * @param {unknown} value
+ */
+const toJSON = (value) => {
+  switch (value) {
+    case Infinity:
+    case -Infinity:
+      return undefined
+    default: {
+      switch (typeof value) {
+        case 'function':
+        case 'undefined':
+        case 'symbol':
+          return undefined
+        case 'number':
+          return isNaN(/** @type {number} */ (value)) ? undefined : value
+        case 'object': {
+          if (
+            typeof (/** @type {{toJSON?:unknown}} */ (value)?.toJSON) ==
+            'function'
+          ) {
+            return /** @type {{toJSON?:unknown}} */ (value).toJSON()
```

Comment:
should this maybe be `toJSON(value.toJSON())` to normalize whatever that function returns?